x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/shards/ShardsCollector.java

```diff
@@ -46,28 +48,68 @@ protected Collection<MonitoringDoc> doCollect(final MonitoringDoc.Node node, fin
-                            MonitoringDoc.Node shardNode = null;
-                            if (shard.assignedToNode()) {
+                for (String index : indices) {
+                    IndexRoutingTable indexRoutingTable = routingTable.index(index);
+                    if (indexRoutingTable != null) {
+                        final int shardCount = indexRoutingTable.size();
+                        for (int i = 0; i < shardCount; i++) {
+                            IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(i);
+
+                            ShardRouting primary = shardRoutingTable.primaryShard();
+                            MonitoringDoc.Node primaryShardNode = null;
+                            if (primary.assignedToNode()) {
                                 // If the shard is assigned to a node, the shard monitoring document refers to this node
-                                shardNode = convertNode(node.getTimestamp(), clusterState.getNodes().get(shard.currentNodeId()));
+                                primaryShardNode = convertNode(node.getTimestamp(), clusterState.getNodes().get(primary.currentNodeId()));
+                            }
+                            results.add(new ShardMonitoringDoc(clusterUuid, timestamp, interval, primaryShardNode, primary, stateUUID, 0));
+
+                            List<ShardRouting> replicas = shardRoutingTable.replicaShards();
+                            for (int j = 0; j < replicas.size(); j++) {
+                                ShardRouting replica = replicas.get(j);
+
+                                MonitoringDoc.Node replicaShardNode = null;
+                                if (replica.assignedToNode()) {
+                                    replicaShardNode = convertNode(
+                                        node.getTimestamp(),
+                                        clusterState.getNodes().get(replica.currentNodeId())
+                                    );
+                                }
+                                results.add(
+                                    new ShardMonitoringDoc(clusterUuid, timestamp, interval, replicaShardNode, replica, stateUUID, j + 1)
+                                );
                             }
-                            results.add(new ShardMonitoringDoc(clusterUuid, timestamp, interval, shardNode, shard, stateUUID));
                         }
                     }
                 }
             }
         }
         return Collections.unmodifiableCollection(results);
     }
+
+    private String[] expandIndexPattern(String[] indicesToMonitor, String[] indices) {
+        final List<String> expandedIndices = new ArrayList<>();
+
+        for (String indexOrPattern : indicesToMonitor) {
+            if (indexOrPattern.contains("*")) {
+                for (String index : indices) {
+                    if (Regex.simpleMatch(indexOrPattern, index)) {
```

Comment:
I've changed it to use Set but I can't think of a way to avoid looping over each index **per** each pattern. If there is a pattern we need to check if any of the indices in the routing table match that pattern, I can't think of how we'd make that a non linear operation.

Reply:
Switching to Set works perfectly fine. The ask about changing looping logic was more about the logic inside of the loop to avoid double-adding indices to the container. Agreed on linear being about the best we can do 😄